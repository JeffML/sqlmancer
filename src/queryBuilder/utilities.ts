import { QueryBuilderContext } from './types'

export function getAlias(name: string, context: QueryBuilderContext): string {
  const index = name.substring(0, 1).toLowerCase()
  if (!context.alias[index]) {
    context.alias[index] = 1
  } else {
    context.alias[index] = context.alias[index]! + 1
  }
  return `${index}${context.alias[index]}`
}

export function getDirection(enumValue: string): string {
  const enumValues: Record<string, string> = {
    ASC: 'asc',
    ASC_NULLS_LAST: 'asc nulls last',
    DESC: 'desc',
    DESC_NULLS_LAST: 'desc nulls last',
  }

  if (!enumValues[enumValue]) {
    throw new Error(`Invalid direction value: ${JSON.stringify(enumValue)}`)
  }

  return enumValues[enumValue]
}

export function getAggregateFunction(key: string): string {
  const aggregates: Record<string, string> = {
    avg: 'avg',
    max: 'max',
    min: 'min',
    sum: 'sum',
  }

  if (!aggregates[key]) {
    throw new Error(`Invalid aggregate function name: ${JSON.stringify(key)}`)
  }

  return aggregates[key]
}

export function getComparisonOperator(key: string): string {
  const operatorMap: Record<string, string> = {
    equal: '=',
    notEqual: '<>',
    in: 'IN',
    notIn: 'NOT IN',
    contains: '@>',
    containedBy: '<@',
    overlaps: '&&',
    greaterThan: '>',
    greaterThanOrEqual: '>=',
    lessThan: '<',
    lessThanOrEqual: '<=',
    like: 'LIKE',
    notLike: 'NOT LIKE',
    iLike: 'ILIKE',
    notILike: 'NOT ILIKE',
    hasKey: '?',
    hasAnyKeys: '?|',
    hasAllKeys: '?&',
  }
  const operator = operatorMap[key]

  if (!operator) {
    throw new Error(`Invalid operator "${key}"`)
  }

  return operator
}
