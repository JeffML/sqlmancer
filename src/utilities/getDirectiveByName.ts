import { GraphQLNamedType } from 'graphql'

export function getDirectiveByName(type: GraphQLNamedType | null | undefined, name: string) {
  return type && type.astNode && type.astNode.directives!.find(directive => directive.name.value === name)
}
