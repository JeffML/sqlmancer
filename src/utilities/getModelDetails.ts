import { isObjectType, DirectiveNode, GraphQLCompositeType, GraphQLOutputType, GraphQLSchema } from 'graphql'
import { parseDirectiveArguments } from './parseDirectiveArguments'

export type ModelDetails = {
  fields: { fieldName: string; columnName?: string; type: GraphQLOutputType; hasDefault: boolean }[]
  joins: { fieldName: string; type: GraphQLOutputType; on: { from: string; to: string }[]; through?: string }[]
  dependencies: { fieldName: string; columns: string[] }[]
}

export function getModelDetails(type: GraphQLCompositeType, schema: GraphQLSchema) {
  const implementingTypes = isObjectType(type) ? [type] : schema.getPossibleTypes(type)
  return implementingTypes.reduce(
    (acc, implementingType) => {
      const fieldMap = implementingType.getFields()
      const fieldNames = Object.keys(fieldMap)
      fieldNames.forEach(fieldName => {
        const field = fieldMap[fieldName]
        const directives = field.astNode!.directives!
        const columnArgs = parseDirectiveArguments(getDirectiveNode('col', directives), schema)
        const joinArgs = parseDirectiveArguments(getDirectiveNode('join', directives), schema)
        const dependArgs = parseDirectiveArguments(getDirectiveNode('depend', directives), schema)
        const ignoreDirective = getDirectiveNode('ignore', directives)
        const hasDefaultDirective = getDirectiveNode('hasDefault', directives)
        if (!ignoreDirective && !joinArgs && !dependArgs) {
          acc.fields.push({
            fieldName,
            type: field.type,
            columnName: columnArgs ? columnArgs.name : undefined,
            hasDefault: !!hasDefaultDirective,
          })
        } else if (joinArgs) {
          acc.joins.push({ fieldName, type: field.type, on: joinArgs.on, through: joinArgs.through })
        } else if (dependArgs) {
          acc.dependencies.push({ fieldName, columns: dependArgs.on })
        }
      })
      return acc
    },
    { fields: [], joins: [], dependencies: [] } as ModelDetails
  )
}

function getDirectiveNode(name: string, directives: ReadonlyArray<DirectiveNode>) {
  const directive = directives.find(directive => directive.name.value === name)
  return directive || null
}
