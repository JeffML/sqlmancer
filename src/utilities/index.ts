export * from './getDirectiveByName'
export * from './getModelDetails'
export * from './getScalarTSType'
export * from './getSqlmancerConfig'
export * from './makeNullable'
export * from './parseDirectiveArguments'
export * from './parseResolveInfo'
export * from './unwrap'
