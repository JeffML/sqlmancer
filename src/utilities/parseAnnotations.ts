import JSON5 from 'json5'
import { parse } from 'doctrine'
import { annotations, AnnotationMap } from '../annotations'

export function parseAnnotations(originalDescription: string | undefined | null) {
  const { description, tags } = parse(originalDescription || '', { tags: Object.keys(annotations) })
  return {
    description,
    annotations: tags.reduce((acc, tag) => {
      const name = tag.title as keyof typeof annotations
      const Annotation = annotations[name] as any
      try {
        acc[name] = new Annotation(JSON5.parse(tag.description || 'null'))
      } catch (error) {
        error.message = `Invalid argument provided for annotation @${name}.\n${error.message}`
        throw error
      }
      return acc
    }, {} as AnnotationMap),
  }
}
