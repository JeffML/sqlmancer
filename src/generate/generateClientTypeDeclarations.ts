import { isEnumType, GraphQLEnumType, GraphQLSchema, isNonNullType } from 'graphql'
import { Writable } from 'stream'

import { getSqlmancerConfig, unwrap } from '../utilities'

export function generateClientTypeDeclarations(schema: GraphQLSchema, stream: Writable): void {
  const { dialect, models } = getSqlmancerConfig(schema)

  stream.write(`
import Knex from 'knex';
import {
  AggregateBuilder,
  BuilderOptions,
  CreateManyBuilder,
  CreateOneBuilder,
  DeleteByIdBuilder,
  DeleteManyBuilder,
  FindByIdBuilder,
  FindManyBuilder,
  FindOneBuilder,
  UpdateByIdBuilder,
  UpdateManyBuilder
} from 'sqlmancer';

export declare type ID = number | string;
export declare type JSON = boolean | number | string | null | JSONArray | JSONObject;
export interface JSONObject {
  [key: string]: JSON;
}
export declare type JSONArray = Array<JSON>;
`)

  Object.keys(models).forEach(name => {
    const { primaryKey, fields, associations } = models[name]

    stream.write(`
export declare type ${name}Fields = {\n${Object.keys(fields)
      .map(fieldName => `  ${fieldName}: ${getOutputFieldType(fields[fieldName].mappedType)};`)
      .join('\n')}
}
    `)

    const idFields = Object.keys(fields).filter(fieldName => fields[fieldName].mappedType === 'ID')
    stream.write(`
export declare type ${name}Ids = ${
      idFields.length ? idFields.map(fieldName => `'${fieldName}'`).join(' | ') : 'unknown'
    };
    `)

    const enums = Object.keys(fields).reduce((acc, fieldName) => {
      const field = fields[fieldName]
      const unwrappedType = unwrap(field.type)
      if (isEnumType(unwrappedType)) {
        acc[unwrappedType.name] = unwrappedType
      }
      return acc
    }, {} as Record<string, GraphQLEnumType>)
    stream.write(`
export declare type ${name}Enums = ${Object.keys(enums).length ? Object.keys(enums).join(' | ') : 'unknown'};
    `)

    stream.write(`
export declare type ${name}Associations = {\n${Object.keys(associations)
      .map(
        name =>
          `  ${name}: [${associations[name].modelName}Find${associations[name].isMany ? 'Many' : 'One'}Builder, ${
            associations[name].modelName
          }AggregateBuilder];`
      )
      .join('\n')}
}
    `)

    stream.write(`
export declare type ${name}CreateFields = {\n${Object.keys(fields)
      .map(fieldName => {
        const field = fields[fieldName]
        const required = isNonNullType(field.type) && !field.hasDefault
        return `  ${fieldName}${required ? '' : '?'}: ${getInputFieldType(field.mappedType)};`
      })
      .join('\n')}
};
    `)

    stream.write(`
export declare type ${name}UpdateFields = {\n${Object.keys(fields)
      .filter(fieldName => fields[fieldName].column !== primaryKey)
      .map(fieldName => `  ${fieldName}?: ${getInputFieldType(fields[fieldName].mappedType)};`)
      .join('\n')}
};
    `)

    Object.keys(enums).forEach(enumName => {
      const enumType = enums[enumName]
      stream.write(`
export declare enum ${enumName} {\n${enumType
        .getValues()
        .map(enumValue => `  ${enumValue.name} = ${JSON.stringify(enumValue.value)},`)
        .join('\n')}
}`)
    })

    stream.write(
      `
export declare class ${name}FindOneBuilder<TSelected extends Pick<${name}Fields, any> = ${name}Fields> extends FindOneBuilder<
  '${dialect}',
  ${name}Fields,
  ${name}Ids,
  ${name}Enums,
  ${name}Associations,
  TSelected
> {
  constructor(options: BuilderOptions);
}

export declare class ${name}FindManyBuilder<TSelected extends Pick<${name}Fields, any> = ${name}Fields> extends FindManyBuilder<
  '${dialect}',
  ${name}Fields,
  ${name}Ids,
  ${name}Enums,
  ${name}Associations,
  TSelected
> {
  constructor(options: BuilderOptions);
}

export declare class ${name}FindByIdBuilder<TSelected extends Pick<${name}Fields, any> = ${name}Fields> extends FindByIdBuilder<
  ${name}Fields,
  ${name}Ids,
  ${name}Enums,
  ${name}Associations,
  TSelected
> {
  constructor(options: BuilderOptions, pk: ID);
}

export declare class ${name}AggregateBuilder<
  TSelected extends Pick<${name}Fields, any> = ${name}Fields
> extends AggregateBuilder<'postgres', ${name}Fields, ${name}Ids, ${name}Enums, ${name}Associations> {
  constructor(options: BuilderOptions);
}
      `
    )

    stream.write(`
export declare class ${name}DeleteManyBuilder extends DeleteManyBuilder<
  '${dialect}',
  ${name}Fields,
  ${name}Ids,
  ${name}Enums,
  ${name}Associations
> {
  constructor(options: BuilderOptions);
}

export declare class ${name}DeleteByIdBuilder extends DeleteByIdBuilder {
  constructor(options: BuilderOptions, pk: ID);
}

export declare class ${name}CreateManyBuilder extends CreateManyBuilder<${name}CreateFields> {
  constructor(options: BuilderOptions, data: ${name}CreateFields[]);
}

export declare class ${name}CreateOneBuilder extends CreateOneBuilder<${name}CreateFields> {
  constructor(options: BuilderOptions, data: ${name}CreateFields);
}

export declare class ${name}UpdateManyBuilder extends UpdateManyBuilder<
  '${dialect}',
  ${name}UpdateFields,
  ${name}Fields,
  ${name}Ids,
  ${name}Enums,
  ${name}Associations
> {
  constructor(options: BuilderOptions, data: ${name}UpdateFields);
}

export declare class ${name}UpdateByIdBuilder extends UpdateByIdBuilder<${name}UpdateFields> {
  constructor(options: BuilderOptions, pk: ID, data: ${name}UpdateFields);
}
  `)
  })

  stream.write(`
type SqlmancerClient = Knex & {
  models: {${Object.keys(models)
    .map(name => {
      const { readOnly } = models[name]
      return `
    ${name}: {
      findById: (id: ID) => ${name}FindByIdBuilder;
      findMany: () => ${name}FindManyBuilder;
      findOne: () => ${name}FindOneBuilder;
      aggregate: () => ${name}AggregateBuilder;${
        readOnly
          ? ''
          : `
      createMany: (input: ${name}CreateFields[]) => ${name}CreateManyBuilder;
      createOne: (input: ${name}CreateFields) => ${name}CreateOneBuilder;
      deleteById: (id: ID) => ${name}DeleteByIdBuilder;
      deleteMany: () => ${name}DeleteManyBuilder;
      updateById: (id: ID, input: ${name}UpdateFields) => ${name}UpdateByIdBuilder;
      updateMany: (input: ${name}UpdateFields) => ${name}UpdateManyBuilder;`
      }
    };`
    })
    .join('')}
  };
};
  `)
}

function getOutputFieldType(mappedType: string) {
  switch (mappedType) {
    case 'Date':
      return 'string'
    case 'Date[]':
      return 'string[]'
    default:
      return mappedType
  }
}

function getInputFieldType(mappedType: string) {
  switch (mappedType) {
    case 'Date':
      return 'Date | string'
    case 'Date[]':
      return 'Date[] | string[]'
    default:
      return mappedType
  }
}
