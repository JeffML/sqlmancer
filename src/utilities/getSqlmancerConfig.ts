import _ from 'lodash'
import { GraphQLSchema } from 'graphql'

import { SqlmancerConfig } from '../types'
import { getDirectiveByName } from './getDirectiveByName'
import { parseDirectiveArguments } from './parseDirectiveArguments'

let config: SqlmancerConfig

export function getSqlmancerConfig(schema: GraphQLSchema): SqlmancerConfig {
  if (!config) {
    const sqlmancerDirective = getDirectiveByName(schema.getQueryType(), 'sqlmancer')

    if (!sqlmancerDirective) {
      throw new Error(
        'Unable to parse Sqlmancer configuration from type definitions. Did you include the @sqlmancer directive on your Query type?'
      )
    }

    const args = parseDirectiveArguments(sqlmancerDirective, schema)!
    config = {
      ...args.config,
      dialect: _.lowerCase(args.config.dialect),
    }
  }

  return config
}
