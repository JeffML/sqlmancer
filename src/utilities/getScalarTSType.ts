import _ from 'lodash'
import { GraphQLSchema } from 'graphql'
import { getSqlmancerConfig } from './getSqlmancerConfig'

type PossibleScalarTSTypes = 'string' | 'number' | 'boolean' | 'JSON' | 'ID'

let scalarMap: Record<string, PossibleScalarTSTypes>

export function getScalarTSType(schema: GraphQLSchema, name: string) {
  return getScalarMap(schema)[name]
}

function getScalarMap(schema: GraphQLSchema) {
  if (!scalarMap) {
    const config = getSqlmancerConfig(schema)
    const baseScalars = {
      string: ['String'],
      number: ['Int', 'Float'],
      boolean: ['Boolean'],
      JSON: [],
      ID: ['ID'],
    }
    const scalarsByTSType = _.mergeWith({}, baseScalars, config.customScalars, (obj, src) => {
      if (Array.isArray(obj)) {
        return obj.concat(src)
      }
    }) as Record<string, string[]>
    scalarMap = Object.keys(scalarsByTSType).reduce((acc, tsType) => {
      if (tsType === 'string' || tsType === 'number' || tsType === 'boolean' || tsType === 'JSON' || tsType === 'ID') {
        scalarsByTSType[tsType].forEach(scalarName => {
          acc[scalarName] = tsType
        })
      }
      return acc
    }, {} as Record<string, PossibleScalarTSTypes>)
  }
  return scalarMap
}
