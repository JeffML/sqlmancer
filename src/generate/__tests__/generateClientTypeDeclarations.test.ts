import { BufferWritableMock } from 'stream-mock'
import { generateClientTypeDeclarations } from '../generateClientTypeDeclarations'
import { withDialects } from '../../__tests__/utilities'

describe('generateClientTypeDeclarations', () => {
  withDialects((_client, _rollback, typeDefs) => {
    test('correct usage', async () => {
      const stream = new BufferWritableMock()
      generateClientTypeDeclarations(typeDefs, stream)
      stream.end()
      await new Promise((resolve, reject) => {
        stream.on('finish', () => {
          expect((stream.flatData as Buffer).toString()).toMatchSnapshot()
          resolve()
        })
        stream.on('error', reject)
      })
    })
  })
})
