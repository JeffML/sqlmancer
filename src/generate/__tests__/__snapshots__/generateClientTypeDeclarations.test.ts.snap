// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateClientTypeDeclarations correct usage 1`] = `
"import Knex from 'knex';
import {
  AggregateBuilder,
  CreateManyBuilder,
  CreateOneBuilder,
  DeleteByIdBuilder,
  DeleteManyBuilder,
  FindByIdBuilder,
  FindManyBuilder,
  FindOneBuilder,
  UpdateByIdBuilder,
  UpdateManyBuilder
} from 'sqlmancer';

export type ID = number | string;
export type JSON = boolean | number | string | null | JSONArray | JSONObject;
export interface JSONObject {
  [key: string]: JSON;
}
export type JSONArray = Array<JSON>;

export type ActorFields = {
  id: ID;
  firstName: string;
  lastName: string;
}
    
export type ActorIds = 'id';
    
export type ActorEnums = unknown;
    
export type ActorAssociations = {
  films: [FilmFindManyBuilder, FilmAggregateBuilder];
}
    
export type ActorCreateFields = {
  id: ID;
  firstName: string;
  lastName: string;
};
    
export type ActorUpdateFields = {
  firstName?: string;
  lastName?: string;
};
    
export type ActorFindOneBuilder<TSelected extends Pick<ActorFields, any> = ActorFields> = FindOneBuilder<
  'postgres',
  ActorFields,
  ActorIds,
  ActorEnums,
  ActorAssociations,
  TSelected
>

export type ActorFindManyBuilder<TSelected extends Pick<ActorFields, any> = ActorFields> = FindManyBuilder<
  'postgres',
  ActorFields,
  ActorIds,
  ActorEnums,
  ActorAssociations,
  TSelected
>

export type ActorFindByIdBuilder<TSelected extends Pick<ActorFields, any> = ActorFields> = FindByIdBuilder<
  ActorFields,
  ActorIds,
  ActorEnums,
  ActorAssociations,
  TSelected
>

export type ActorAggregateBuilder = AggregateBuilder<'postgres', ActorFields, ActorIds, ActorEnums, ActorAssociations>
      
export type ActorDeleteManyBuilder = DeleteManyBuilder<
  'postgres',
  ActorFields,
  ActorIds,
  ActorEnums,
  ActorAssociations
>

export type ActorDeleteByIdBuilder = DeleteByIdBuilder

export type ActorCreateManyBuilder = CreateManyBuilder<ActorCreateFields>

export type ActorCreateOneBuilder = CreateOneBuilder<ActorCreateFields>

export type ActorUpdateManyBuilder = UpdateManyBuilder<
  'postgres',
  ActorUpdateFields,
  ActorFields,
  ActorIds,
  ActorEnums,
  ActorAssociations
>

export type ActorUpdateByIdBuilder = UpdateByIdBuilder<ActorUpdateFields>
  
export type FilmFields = {
  id: ID;
  title: string;
  releaseYear: number;
}
    
export type FilmIds = 'id';
    
export type FilmEnums = unknown;
    
export type FilmAssociations = {
  actors: [ActorFindManyBuilder, ActorAggregateBuilder];
}
    
export type FilmCreateFields = {
  id: ID;
  title: string;
  releaseYear: number;
};
    
export type FilmUpdateFields = {
  title?: string;
  releaseYear?: number;
};
    
export type FilmFindOneBuilder<TSelected extends Pick<FilmFields, any> = FilmFields> = FindOneBuilder<
  'postgres',
  FilmFields,
  FilmIds,
  FilmEnums,
  FilmAssociations,
  TSelected
>

export type FilmFindManyBuilder<TSelected extends Pick<FilmFields, any> = FilmFields> = FindManyBuilder<
  'postgres',
  FilmFields,
  FilmIds,
  FilmEnums,
  FilmAssociations,
  TSelected
>

export type FilmFindByIdBuilder<TSelected extends Pick<FilmFields, any> = FilmFields> = FindByIdBuilder<
  FilmFields,
  FilmIds,
  FilmEnums,
  FilmAssociations,
  TSelected
>

export type FilmAggregateBuilder = AggregateBuilder<'postgres', FilmFields, FilmIds, FilmEnums, FilmAssociations>
      
export type FilmDeleteManyBuilder = DeleteManyBuilder<
  'postgres',
  FilmFields,
  FilmIds,
  FilmEnums,
  FilmAssociations
>

export type FilmDeleteByIdBuilder = DeleteByIdBuilder

export type FilmCreateManyBuilder = CreateManyBuilder<FilmCreateFields>

export type FilmCreateOneBuilder = CreateOneBuilder<FilmCreateFields>

export type FilmUpdateManyBuilder = UpdateManyBuilder<
  'postgres',
  FilmUpdateFields,
  FilmFields,
  FilmIds,
  FilmEnums,
  FilmAssociations
>

export type FilmUpdateByIdBuilder = UpdateByIdBuilder<FilmUpdateFields>
  
export type SqlmancerClient = Knex & {
  models: {
    Actor: {
      findById: (id: ID) => ActorFindByIdBuilder;
      findMany: () => ActorFindManyBuilder;
      findOne: () => ActorFindOneBuilder;
      aggregate: () => ActorAggregateBuilder;
      createMany: (input: ActorCreateFields[]) => ActorCreateManyBuilder;
      createOne: (input: ActorCreateFields) => ActorCreateOneBuilder;
      deleteById: (id: ID) => ActorDeleteByIdBuilder;
      deleteMany: () => ActorDeleteManyBuilder;
      updateById: (id: ID, input: ActorUpdateFields) => ActorUpdateByIdBuilder;
      updateMany: (input: ActorUpdateFields) => ActorUpdateManyBuilder;
    };
    Film: {
      findById: (id: ID) => FilmFindByIdBuilder;
      findMany: () => FilmFindManyBuilder;
      findOne: () => FilmFindOneBuilder;
      aggregate: () => FilmAggregateBuilder;
      createMany: (input: FilmCreateFields[]) => FilmCreateManyBuilder;
      createOne: (input: FilmCreateFields) => FilmCreateOneBuilder;
      deleteById: (id: ID) => FilmDeleteByIdBuilder;
      deleteMany: () => FilmDeleteManyBuilder;
      updateById: (id: ID, input: FilmUpdateFields) => FilmUpdateByIdBuilder;
      updateMany: (input: FilmUpdateFields) => FilmUpdateManyBuilder;
    };
  };
};
  "
`;
